Source/DafnyCore/AST/IndentationFormatter.cs

```diff
@@ -0,0 +1,1875 @@
+        //
+        // const x := 2;
+        //   /* Start of comment
+        // end of comment */
+        //
+        while (tryAgain)
+        // decreases newAbsoluteIndent - relativeIndent
+        {
+          var canIndentLinesStartingWithStar = true;
+          tryAgain = false;
+          result = new Regex($@"(?<=\r?\n|\r(?!\n))(?<existingIndent>[ \t]*)(?<star>\*)?").Replace(originalComment, match1 => {
+            if (canIndentLinesStartingWithStar && match1.Groups["star"].Success) {
+              return indentationBefore + (doubleStar ? "  *" : " *");
+            }
+
+            canIndentLinesStartingWithStar = false;
+            // Reindent in block:
+            var newIndent = match1.Groups["existingIndent"].Value.Length + relativeIndent;
+            if (newIndent < 0) {
+              relativeIndent -= newIndent;
+              tryAgain = true;
+              newIndent = 0;
+            }
+
+            return new string(' ', newIndent) + (match1.Groups["star"].Success ? match1.Groups["star"].Value : "");
+          });
+        }
+
+        previousMatchWasSingleLineCommentToAlign = false;
+        if (precededByNewline) {
+          return new string(' ', absoluteOriginalIndent + relativeIndent) + result;
+        }
+
+        return new string(' ', match.Groups["currentIndent"].Length + relativeIndent - initialRelativeIndent) + result;
+      }
+
+      if (commentType.StartsWith("//")) {
+        if (commentType.StartsWith("///") && !commentType.StartsWith("////")) {
+          // No indentation
+          return match.Groups["commentType"].Value;
+        }
+
+        if (previousMatchWasSingleLineCommentToAlign) {
+          if (originalCommentIndent == match.Groups["currentIndent"].Value.Length) {
+            return new string(' ', newCommentIndent) + match.Groups["commentType"].Value;
+          }
+        }
+
+        var referenceToken = token.Next;
+        if (match.Groups["caseCommented"].Success && token.Next != null && (token.Next.val == match.Groups["caseCommented"].Value || FirstTokenOnLineIs(token, t => {
+          referenceToken = t;
+          return t.val == match.Groups["caseCommented"].Value;
+        }))) {
+          indentationBefore = new string(' ', GetNewTokenVisualIndent(referenceToken, indentationBefore.Length));
+        }
+
+        previousMatchWasSingleLineCommentToAlign = false;
+
+        return indentationBefore + match.Groups["commentType"].Value;
+      }
+
+      if (commentType.StartsWith("\r") || commentType.StartsWith("\n")) {
+        previousMatchWasSingleLineCommentToAlign = false;
+        return (HelperString.BlankNewlinesWithoutSpaces ? "" : indentationBefore) + match.Groups["commentType"].Value;
+      }
+
+      previousMatchWasSingleLineCommentToAlign = false;
+
+      // Last line
+      return lastIndentation;
+    }
+
+    return HelperString.NewlineRegex.Replace(input, ApplyIndentRules);
```

Comment:
> Handling the replace seems like a good job for a lambda instead of a named method, because the handling is tied 1-to-1 to the Regex used for the replace.
I inlined the local method.